spec.html

```diff
@@ -448,73 +480,96 @@ <h1>InnerModuleEvaluation( _module_, _stack_, _index_ )</h1>
-              1. Perform ! _module_.ExecuteModule(_capability_).
-            1. Otherwise,
-              1. Let _result_ be _module_.ExecuteModule().
-              1. If _result_ is a normal completion, then
-                1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, &laquo;*undefined*&raquo;).
-              1. Otherwise,
-                1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo;_result_.[[Value]]&raquo;).
-        1. Let _onFulfilled_ be CreateBuiltinFunction(_stepsFulfilled_).
-        1. Let _stepsReject_ be the following steps with argument _arg_
-          1. Perform ! Call(_capability_.[[Reject]], *undefined*, &laquo;_arg_&raquo;).
-        1. Let _onReject_ be CreateBuiltinFunction(_stepsReject_).
-        1. Perform ! PerformPromiseThen(_promise_, _onFulfilled_, _onRejected_).
-      1. Return.
-    </emu-alg>
+    <emu-clause id="sec-executionfulfilled" aoid="AsyncExecutionFulfilled">
+      <h1><ins>AsyncExecutionFulfilled ()</ins></h1>
+      <emu-alg>
+        1. Let _f_ be the active function object.
+        1. Let _module_ be _f_.[[Module]].
+        1. Assert: _module_.[[Status]] is `"evaluating-async"`.
+        1. Set _module_.[[Status]] to `"evaluated"`.
+        1. For each Module _m_ of _module_.[[AsyncParentModules]], do
+          1. Decrement _m_.[[PendingAsyncDependencies]] by 1.
+          1. If _m_.[[PendingAsyncDependencies]] is 0 and _m_.[[EvaluationError]] is *undefined*, then
+            1. Assert _m_.[[Status]] is `"evaluating-async"`.
+            1. Perform ! _module_.ExecuteModule().
+        1. If _module_.[[DFSAncestorIndex]] equals _module_.[[DFSIndex]], then
+          1. Perform ! Call(_module_.[[EvaluationCapability]].[[Resolve]], *undefined*, &laquo;*undefined*&raquo;).
+        1. Return *undefined*.
+      </emu-alg>
+      <emu-note>
+        <p>When propagating ExecuteModule calls up the parent stack through multiple AsyncExecutionFulfilled handlers, the execution context should not be changed apart from the direct execution of the modules being executed in the stack.</p>
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-executionrejected" aoid="AsyncExecutionRejected">
+      <h1><ins>AsyncExecutionRejected ( _error_ )</ins></h1>
+      <emu-alg>
+        1. Let _f be the active function object.
+        1. Let _module_ be _f_.[[Module]].
+        1. Assert: _module_.[[Status]] is `"evaluating-async"`.
+        1. Assert: _module_.[[EvaluationError]] is *undefined*.
+        1. Set _module_.[[Status]] to `"evaluated"`.
+        1. Set _module_.[[EvaluationError]] to ThrowCompletion(_error_).
+        1. For each Module _m_ of _module_.[[AsyncParentModules]], do
+          1. If _m_.[[Status]] is `"evaluating-async"`, then
+            1. Let _stepsRejected_ be the steps of _AsyncExecutionRejected_.
+            1. Let _parentRejected_ be CreateBuiltinFunction(_stepsRejected_, &laquo; [[Module]] &raquo;).
```

Comment:
Why create a function to call it immediately? Would it be better to factor out an abstract op instead, or try to define this iteratively rather than recursively?

Reply:
Yes this needs a new approach, I just didn't know what was best.

AsyncExectionResolved and AsyncExecutionRejected are global handlers that take a single module argument, and are designed to support being passed in to promise resolution.

But they should also support being called natively.

The way they are written to read the `_module_` argument was for the promises builtin cases, but then I didn't know how to call them in a way that could work for both cases.

Perhaps they should instead be wrapped in the promise cases, instead of making a builtin function here I guess?

Reply:
I've fixed this up in the latest commit.